File: gravy/portfolio_helpers_test.go
package gravy

import (
	"testing"

	dailyprices_pb "github.com/d-sparks/gravy/data/dailyprices/proto"
	supervisor_pb "github.com/d-sparks/gravy/supervisor/proto"
	"github.com/stretchr/testify/assert"
)

var (
	algorithmID = &supervisor_pb.AlgorithmId{AlgorithmId: "test"}

	portfolio = &supervisor_pb.Portfolio{Stocks: map[string]float64{"MSFT": 3.0, "GOOG": 2.0}, Usd: 1.0}

	prices = &dailyprices_pb.DailyPrices{
		StockPrices: map[string]*dailyprices_pb.DailyPrices_StockPrices{
			"MSFT": &dailyprices_pb.DailyPrices_StockPrices{Close: 5.0},
			"GOOG": &dailyprices_pb.DailyPrices_StockPrices{Close: 10.0},
			"FB":   &dailyprices_pb.DailyPrices_StockPrices{Close: 7.0},
		},
	}
)

// TestPortfolioValue tests the basic arithmetic of portfolio value.
func TestPortfolioValue(t *testing.T) {
	assert.Equal(t, 36.0, PortfolioValue(portfolio, prices))
}

// TestTargetUniformInvestment tests the arithmetic. We expect portfolio value / # stocks.
func TestTargetUniformInvestment(t *testing.T) {
	assert.Equal(t, 12.0, TargetUniformInvestment(portfolio, prices))
}

// TestSellEverythingWithStop tests with custom stop lambda.
func TestSellEverythingWithStop(t *testing.T) {
	stop := func(ticker string) float64 {
		return map[string]float64{"MSFT": 0.1, "GOOG": 0.2}[ticker]
	}
	orders := SellEverythingWithStop(algorithmID, portfolio, stop)

	expected := map[string]*supervisor_pb.Order{
		"MSFT": &supervisor_pb.Order{
			AlgorithmId: algorithmID, Ticker: "MSFT", Volume: -3.0, Stop: 0.1, Limit: 0.0,
		},
		"GOOG": &supervisor_pb.Order{
			AlgorithmId: algorithmID, Ticker: "GOOG", Volume: -2.0, Stop: 0.2, Limit: 0.0,
		},
	}
	for _, order := range orders {
		assert.Equal(t, expected[order.GetTicker()], order)
	}
}

// TestSellEverythingWithStopPercent tests with fixed stop percent.
func TestSellEverythingWithStopPercent(t *testing.T) {
	orders := SellEverythingWithStopPercent(algorithmID, portfolio, prices, 0.5)

	expected := map[string]*supervisor_pb.Order{
		"MSFT": &supervisor_pb.Order{
			AlgorithmId: algorithmID, Ticker: "MSFT", Volume: -3.0, Stop: 2.5, Limit: 0.0,
		},
		"GOOG": &supervisor_pb.Order{
			AlgorithmId: algorithmID, Ticker: "GOOG", Volume: -2.0, Stop: 5.0, Limit: 0.0,
		},
	}
	for _, order := range orders {
		assert.Equal(t, expected[order.GetTicker()], order)
	}
}

// TestSellEverythingWithMarketOrder tests with no stop.
func TestSellEverythingMarketOrder(t *testing.T) {
	orders := SellEverythingMarketOrder(algorithmID, portfolio)

	expected := map[string]*supervisor_pb.Order{
		"MSFT": &supervisor_pb.Order{
			AlgorithmId: algorithmID, Ticker: "MSFT", Volume: -3.0, Stop: 0.0, Limit: 0.0,
		},
		"GOOG": &supervisor_pb.Order{
			AlgorithmId: algorithmID, Ticker: "GOOG", Volume: -2.0, Stop: 0.0, Limit: 0.0,
		},
	}
	for _, order := range orders {
		assert.Equal(t, expected[order.GetTicker()], order)
	}
}

var testAlgorithmID = &supervisor_pb.AlgorithmId{AlgorithmId: "TEST"}

// TestInvestApproximatelyUniformlyHappyPath is the happy path test case with a single stock.
func TestInvestApproximatelyUniformlyHappyPath(t *testing.T) {
	var dailyPrices dailyprices_pb.DailyPrices
	var portfolio supervisor_pb.Portfolio

	dailyPrices.StockPrices = map[string]*dailyprices_pb.DailyPrices_StockPrices{
		"MSFT": &dailyprices_pb.DailyPrices_StockPrices{Close: 10.0},
	}
	portfolio.Usd = 1000.0

	orders := InvestApproximatelyUniformly(testAlgorithmID, &portfolio, &dailyPrices)

	assert.Equal(t, 1, len(orders))
	order := orders[0]
	assert.Equal(t, "MSFT", order.GetTicker())
	assert.Equal(t, 99.0, order.GetVolume())
	assert.Equal(t, 10.1, order.GetLimit())
}

// TestInvestApproximatelyUniformlyTwoStocks is the happy path test case with a two stocks and no tradeoff.
func TestInvestApproximatelyUniformlyTwoStocks(t *testing.T) {
	var dailyPrices dailyprices_pb.DailyPrices
	var portfolio supervisor_pb.Portfolio

	dailyPrices.StockPrices = map[string]*dailyprices_pb.DailyPrices_StockPrices{
		"MSFT": &dailyprices_pb.DailyPrices_StockPrices{Close: 10.0},
		"GOOG": &dailyprices_pb.DailyPrices_StockPrices{Close: 20.0},
	}
	portfolio.Usd = 1000.0

	orders := InvestApproximatelyUniformly(testAlgorithmID, &portfolio, &dailyPrices)

	expectedVolume := map[string]float64{"MSFT": 49.0, "GOOG": 25.0}
	expectedLimit := map[string]float64{"MSFT": 10.1, "GOOG": 20.2}
	assert.Equal(t, 3, len(orders))
	volume := map[string]float64{}
	for _, order := range orders {
		assert.Equal(t, expectedLimit[order.GetTicker()], order.GetLimit())
		volume[order.GetTicker()] += order.GetVolume()
	}

	assert.Equal(t, expectedVolume, volume)
}

// TestInvestApproximatelyUniformlyManyStocks tests the case of many stocks.
func TestInvestApproximatelyUniformlyManyStocks(t *testing.T) {
	var dailyPrices dailyprices_pb.DailyPrices
	var portfolio supervisor_pb.Portfolio

	dailyPrices.StockPrices = map[string]*dailyprices_pb.DailyPrices_StockPrices{
		"MSFT": &dailyprices_pb.DailyPrices_StockPrices{Close: 10.0},
		"GOOG": &dailyprices_pb.DailyPrices_StockPrices{Close: 20.0},
		"FB":   &dailyprices_pb.DailyPrices_StockPrices{Close: 7.0},
		"APPL": &dailyprices_pb.DailyPrices_StockPrices{Close: 150.0},
		"NVDA": &dailyprices_pb.DailyPrices_StockPrices{Close: 9.0},
		"GM":   &dailyprices_pb.DailyPrices_StockPrices{Close: 0.50},
		"FORD": &dailyprices_pb.DailyPrices_StockPrices{Close: 1.0},
	}
	portfolio.Usd = 1000.0

	orders := InvestApproximatelyUniformly(testAlgorithmID, &portfolio, &dailyPrices)

	expectedVolume := map[string]float64{
		"APPL": 1.0,
		"FB":   20.0,
		"FORD": 141.0,
		"GM":   282.0,
		"GOOG": 7.0,
		"MSFT": 14.0,
		"NVDA": 15.0,
	}
	expectedLimit := map[string]float64{
		"APPL": 151.5,
		"FB":   7.07,
		"FORD": 1.01,
		"GM":   0.505,
		"GOOG": 20.2,
		"MSFT": 10.1,
		"NVDA": 9.09,
	}

	assert.Equal(t, 7, len(orders))
	volume := map[string]float64{}
	limit := map[string]float64{}
	totalLimit := 0.0
	for _, order := range orders {
		volume[order.GetTicker()] += order.GetVolume()
		limit[order.GetTicker()] = order.GetLimit()
		totalLimit += order.GetLimit() * order.GetVolume()
	}

	assert.Equal(t, expectedVolume, volume)
	assert.Equal(t, expectedLimit, limit)
	assert.Less(t, 996.0, totalLimit)
}

File: gravy/portfolio_helpers_test.py
import gravy.portfolio_helpers as portfolio_helpers
import json
import unittest
import google.protobuf.json_format as json_format

from data.dailyprices.proto import daily_prices_pb2
from supervisor.proto import supervisor_pb2


test_algorithm_id = supervisor_pb2.AlgorithmId(algorithm_id="test")


class TestInvestApproximatelyUniformly(unittest.TestCase):

    def test_many_stocks(self):
        # Inputs
        portfolio = supervisor_pb2.Portfolio(usd=1000.0)
        daily_data = json_format.Parse(json.dumps({
            "prices": {
                "MSFT": {"close": 10.0},
                "GOOG": {"close": 20.0},
                "FB": {"close": 7.0},
                "APPL": {"close": 150.0},
                "NVDA": {"close": 9.0},
                "GM": {"close": 0.50},
                "FORD": {"close": 1.0},
            },
        }), daily_prices_pb2.DailyData())
        # Expected outputs
        expected_volume = {
            "APPL": 1.0,
            "FB": 20.0,
            "FORD": 141.0,
            "GM": 282.0,
            "GOOG": 7.0,
            "MSFT": 14.0,
            "NVDA": 15.0,
        }
        expected_limit = {
            "APPL": 151.5,
            "FB": 7.07,
            "FORD": 1.01,
            "GM": 0.505,
            "GOOG": 20.2,
            "MSFT": 10.1,
            "NVDA": 9.09,
        }
        # Create orders to be tested
        orders = portfolio_helpers.invest_approximately_uniformly(
            test_algorithm_id, portfolio, daily_data)
        volume = dict()
        limit = dict()
        total_limit = 0.0
        for order in orders:
            volume[order.ticker] = order.volume
            limit[order.ticker] = order.limit
            total_limit += order.limit * order.volume
        # Assertions
        self.assertEqual(len(orders), 7)
        self.assertEqual(volume, expected_volume)
        self.assertEqual(limit, expected_limit)
        self.assertLess(996.0, total_limit)

    def test_many_stocks_some_already_held(self):
        # Inputs
        portfolio = json_format.Parse(json.dumps({
            "usd": 664.0,
            "stocks": {
                "MSFT": 7.0,
                "NVDA": 14.0,
                "GOOG": 7.0,
            }
        }), supervisor_pb2.Portfolio())
        daily_data = json_format.Parse(json.dumps({
            "prices": {
                "MSFT": {"close": 10.0},
                "GOOG": {"close": 20.0},
                "FB": {"close": 7.0},
                "APPL": {"close": 150.0},
                "NVDA": {"close": 9.0},
                "GM": {"close": 0.50},
                "FORD": {"close": 1.0},
            },
        }), daily_prices_pb2.DailyData())
        # Expected outputs
        expected_volume = {
            "APPL": 1.0,
            "FB": 20.0,
            "FORD": 141.0,
            "GM": 282.0,
            "MSFT": 7.0,
            "NVDA": 1.0,
        }
        expected_limit = {
            "APPL": 151.5,
            "FB": 7.07,
            "FORD": 1.01,
            "GM": 0.505,
            "MSFT": 10.1,
            "NVDA": 9.09,
        }
        # Create orders to be tested
        orders = portfolio_helpers.invest_approximately_uniformly(
            test_algorithm_id, portfolio, daily_data)
        volume = dict()
        limit = dict()
        total_limit = 0.0
        for order in orders:
            volume[order.ticker] = order.volume
            limit[order.ticker] = order.limit
            total_limit += order.limit * order.volume
        # Assertions
        self.assertEqual(len(orders), 6)
        self.assertEqual(volume, expected_volume)
        self.assertEqual(limit, expected_limit)
        self.assertLess(657.0, total_limit)
        self.assertLess(total_limit, 664.0)

    def test_partial(self):
        # Inputs
        portfolio = json_format.Parse(json.dumps({
            "usd": 100,
            "stocks": {
                "APPL": 1.0,
            }
        }), supervisor_pb2.Portfolio())
        daily_data = json_format.Parse(json.dumps({
            "prices": {
                "MSFT": {"close": 25.0},
                "GOOG": {"close": 25.0},
                "FB": {"close": 2.5},
                "APPL": {"close": 33.0},
            },
        }), daily_prices_pb2.DailyData())
        # Expected outputs
        expected_volume = {
            "FB": 13.0,
            "GOOG": 1.0,
        }
        expected_limit = {
            "FB": 2.525,
            "GOOG": 25.25,
        }
        # Create orders to be tested
        orders = portfolio_helpers.invest_approximately_uniformly_in_targets(
            test_algorithm_id, portfolio, daily_data, ["GOOG", "FB"], 80.0)
        volume = dict()
        limit = dict()
        total_limit = 0.0
        for order in orders:
            if order.ticker not in volume:
                volume[order.ticker] = 0.0
            volume[order.ticker] += order.volume
            limit[order.ticker] = order.limit
            total_limit += order.limit * order.volume
        # Assertions
        self.assertEqual(len(orders), 2)
        self.assertEqual(volume, expected_volume)
        self.assertEqual(limit, expected_limit)
        self.assertLess(58, total_limit)
        self.assertLess(total_limit, 80.0)


class TestOrdersSortedDescending(unittest.TestCase):

    def test_happy_path(self):
        orders = [
            supervisor_pb2.Order(volume=5.0, limit=5.0),
            supervisor_pb2.Order(volume=6.0, limit=8.0),
            supervisor_pb2.Order(volume=7.0, limit=9.0),
            supervisor_pb2.Order(volume=8.0, limit=2.0),
            supervisor_pb2.Order(volume=-8.0, limit=2.0),
            supervisor_pb2.Order(volume=-9.0, limit=3.0),
        ]
        sorted_orders = portfolio_helpers.orders_sorted_descending(orders)
        for i in range(1, len(sorted_orders)):
            prev_order = sorted_orders[i-1]
            order = sorted_orders[i]
            # Sells before buys.
            self.assertFalse(prev_order.volume > 0.0 and order.volume <= 0.0)
            if prev_order.volume <= 0.0:
                continue
            # Buys in descending order.
            self.assertLess(order.limit * order.volume,
                            prev_order.limit * prev_order.volume)


class TestSellOverweightTargetStocks(unittest.TestCase):

    def test_simple(self):
        # Inputs
        portfolio = json_format.Parse(json.dumps({
            "stocks": {
                "MSFT": 6.0,
                "GOOG": 7.0,
                "FB": 6.0,
                "APPL": 7.0,
            },
            "usd": -50.0,
        }), supervisor_pb2.Portfolio())
        daily_data = json_format.Parse(json.dumps({
            "prices": {
                "MSFT": {"close": 10.0},
                "GOOG": {"close": 10.0},
                "FB": {"close": 11.0},
                "APPL": {"close": 8.0},
            },
        }), daily_prices_pb2.DailyData())
        # Expected outputs
        expected_volume = {
            "APPL": -1.0,
            "FB": -1.0,
            "GOOG": -2.0,
            "MSFT": -1.0,
        }
        # Create orders to be tested
        targets = ["GOOG", "FB", "APPL", "MSFT"]
        orders = portfolio_helpers.sell_overweight_target_stocks(
            test_algorithm_id, portfolio, daily_data, targets)
        # Assertions
        for order in orders:
            self.assertEqual(order.volume, expected_volume[order.ticker])
            self.assertEqual(order.stop, 0.99 *
                             daily_data.prices[order.ticker].close)
        self.assertEqual(len(orders), 4)

    def test_restricted(self):
        # Inputs
        portfolio = json_format.Parse(json.dumps({
            "stocks": {
                "MSFT": 6.0,
                "GOOG": 7.0,
                "FB": 6.0,
                "APPL": 7.0,
            },
            "usd": -50.0,
        }), supervisor_pb2.Portfolio())
        daily_data = json_format.Parse(json.dumps({
            "prices": {
                "MSFT": {"close": 10.0},
                "GOOG": {"close": 10.0},
                "FB": {"close": 11.0},
                "APPL": {"close": 8.0},
            },
        }), daily_prices_pb2.DailyData())
        # Expected outputs
        expected_volume = {
            "APPL": -1.0,
            "FB": -1.0,
        }
        # Create orders to be tested
        targets = ["FB", "APPL"]
        orders = portfolio_helpers.sell_overweight_target_stocks(
            test_algorithm_id, portfolio, daily_data, targets)
        # Assertions
        for order in orders:
            self.assertEqual(order.volume, expected_volume[order.ticker])
            self.assertEqual(order.stop, 0.99 *
                             daily_data.prices[order.ticker].close)
        self.assertEqual(len(orders), 2)

    def test_restricted_inverted(self):
        # Inputs
        portfolio = json_format.Parse(json.dumps({
            "stocks": {
                "MSFT": 6.0,
                "GOOG": 7.0,
                "FB": 6.0,
                "APPL": 7.0,
            },
            "usd": -50.0,
        }), supervisor_pb2.Portfolio())
        daily_data = json_format.Parse(json.dumps({
            "prices": {
                "MSFT": {"close": 10.0},
                "GOOG": {"close": 10.0},
                "FB": {"close": 11.0},
                "APPL": {"close": 8.0},
            },
        }), daily_prices_pb2.DailyData())
        # Expected outputs
        expected_volume = {
            "APPL": -1.0,
            "FB": -1.0,
        }
        # Create orders to be tested
        targets = ["GOOG", "MSFT", "BOGUS"]
        orders = portfolio_helpers.sell_overweight_target_stocks(
            test_algorithm_id, portfolio, daily_data, targets,
            invert_targets=True)
        # Assertions
        for order in orders:
            self.assertEqual(order.volume, expected_volume[order.ticker])
            self.assertEqual(order.stop, 0.99 *
                             daily_data.prices[order.ticker].close)
        self.assertEqual(len(orders), 2)

    def test_underweight(self):
        # Inputs
        portfolio = json_format.Parse(json.dumps({
            "stocks": {
                "MSFT": 0.0,
                "GOOG": 7.0,
                "FB": 6.0,
                "APPL": 7.0,
            },
            "usd": 10.0,
        }), supervisor_pb2.Portfolio())
        daily_data = json_format.Parse(json.dumps({
            "prices": {
                "MSFT": {"close": 10.0},
                "GOOG": {"close": 10.0},
                "FB": {"close": 11.0},
                "APPL": {"close": 8.0},
            },
        }), daily_prices_pb2.DailyData())
        # Expected outputs
        expected_volume = {
            "APPL": -1.0,
            "FB": -1.0,
            "GOOG": -2.0,
        }
        # Create orders to be tested
        targets = ["GOOG", "FB", "APPL", "MSFT"]
        orders = portfolio_helpers.sell_overweight_target_stocks(
            test_algorithm_id, portfolio, daily_data, targets)
        # Assertions
        for order in orders:
            self.assertEqual(order.volume, expected_volume[order.ticker])
            self.assertEqual(order.stop, 0.99 *
                             daily_data.prices[order.ticker].close)
        self.assertEqual(len(orders), 3)


if __name__ == '__main__':
    unittest.main()
